src/pressure/VCycleMG.cpp

```diff
@@ -753,154 +612,244 @@ void VCycleMG::Solve(Field &out, Field &tmp, Field const &b, size_t level, bool
+
+    // start/end
+    // inner
+    const size_t start_i = boundary->get_inner_list_level_joined_start(level);
+    const size_t end_i = boundary->get_inner_list_level_joined_end(level) + 1;
+
+    boundary->apply_boundary(data_out, level, type, sync);
+
+    const real rdx2 = 1. / (dx * dx);
+    const real rdy2 = 1. / (dy * dy);
+    const real rdz2 = 1. / (dz * dz);
+
+    // preparation for diffusion step (alpha, beta)
+    const real alphaX = rdx2;
+    const real alphaY = rdy2;
+    const real alphaZ = rdz2;
+
+    const real rbeta = 2. * (alphaX + alphaY + alphaZ);
+    const real beta = 1. / rbeta;
+
+    tmp.copy_data(out);
 #pragma acc data present(out, tmp, b)
-        {
-            size_t it = 0;
-            const size_t max_it = static_cast<size_t>(params->get_int("solver/diffusion/max_iter"));
-            const real tol_res = params->get_real("solver/diffusion/tol_res");
-            real sum;
-            real res = 10000.;
-
-            while (res > tol_res && it < max_it) {
-                ColoredGaussSeidelDiffuse::colored_gauss_seidel_step(out, b, alphaX, alphaY, alphaZ, beta, m_dsign, m_w, sync);
-                boundary->apply_boundary(out.data, level, type, sync); // for res/err only Dirichlet BC
-
-                sum = 0.;
-
-#pragma acc parallel loop independent present(out, tmp, d_iList[:bsize_i]) async
-                for (size_t j = start_i; j < end_i; ++j) {
-                    const size_t i = d_iList[j];
-                    res = b[i] - (rdx2 * (out[i - 1] - 2 * out[i] + out[i + 1])\
- + rdy2 * (out[i - Nx] - 2 * out[i] + out[i + Nx])\
- + rdz2 * (out[i - Nx * Ny] - 2 * out[i] + out[i + Nx * Ny])); //res = rbeta*(out[i] - tmp[i]);
-                    sum += res * res;
-                }
-                // info: in nvvp profile 8byte size copy from to device to/from pageable due to sum!
+    {
+        size_t it = 0;
+        real sum;
+        real res = 10000.;
 
-#pragma acc wait
-                res = sqrt(sum);
-                it++;
+        const size_t neighbour_i = 1;
+        const size_t neighbour_j = Nx;
+        const size_t neighbour_k = Nx * Ny;
+        while (res > m_diffusion_tol_res && it < m_diffusion_max_iter) {
+            JacobiDiffuse::JacobiStep(level, out, tmp, b, alphaX, alphaY, alphaZ, beta, m_dsign, m_w, sync);
+            boundary->apply_boundary(data_out, level, type, sync);
+
+            sum = 0.;
+
+#pragma acc parallel loop independent present(out, tmp, data_inner_list[:bsize_i]) async
+            for (size_t j = start_i; j < end_i; ++j) {
+                const size_t i = data_inner_list[j];
+                res = b[i] - (rdx2 * (out[i - neighbour_i] - 2 * out[i] + data_out[i + neighbour_i])
+                           +  rdy2 * (out[i - neighbour_j] - 2 * out[i] + data_out[i + neighbour_j])
+                           +  rdz2 * (out[i - neighbour_k] - 2 * out[i] + data_out[i + neighbour_k])); //res = rbeta*(data_out[i] - data_tmp[i]);
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247